Plumbing Update

FalseIncarnate opened the pull request:
Sinks, Toilets, and Urinals can now be unanchored and moved with a wrench. (Puddles can't be moved, they are puddles)

![image](https://user-images.githubusercontent.com/8988254/28350398-b3213a4e-6c16-11e7-9bc9-4b0beb3796b1.png)
Here's a kitchen sink moved to the chapel from the old bar.

Standard sinks and toilets can also be rotated when unanchored by hitting them with a wrench again.
- Urinals and kitchen sinks cannot be rotated because they looked horrible in other directional states.
  - I removed the directional states from them because we don't use them (and MAN were they bad).

![image](https://user-images.githubusercontent.com/8988254/28497190-35788962-6f4f-11e7-830a-ea6d72dd6d1c.png)
Here's a sink rotated and anchored to face north in the locker room bathrooms.

When re-anchoring a sink or urinal, it will automagically offset itself to match its direction. This means sinks will appear against walls (or at least the edge of the tile).
- Toilets facing NORTH (this means you see the back of the toilet in the front of your view, not the bowl) now appear IN FRONT of mobs (like chairs!).
- Sinks facing NORTH now also appear IN FRONT of mobs.

While unanchored, sinks cannot be used to wash hands or objects.
- Toilets can still have things hidden in their cistern while unanchored, however.

Showers can now be detached with a welder, though you'll want to turn the shower off first.
- Detached showers become an item you can slap onto walls to reattach them.

Adjusting the temperature on showers with a wrench is now adjusted by toolspeed

Refactored shower code for making and deleting mist (steam) to no longer use spawns and do the following:
- Boiling water now creates mist after 2 seconds, normal water takes the usual 5 seconds
- Freezing water now disperses mist after 7 seconds, if some was present from that shower
- Mist still disperses naturally after 25 seconds if the shower has been turned off.
  - Showers will automatically delete any mist they may have made when destroyed/detached

Also slipped in a new /atom/movable proc to handle transferring fingerprints from one atom to another. Why this wasn't already a thing is a mystery.
- Proc has the option to either overwrite or combine the list of fingerprints on the target atom

Toilets, sinks, urinals, and showers can be purchased in two new cargo crates!
- Urination Station Crate:
  - Costs 10 points, contains 1 "toilet in a box" and 1 "urinal in a box".
- Hygiene Station Crate:
  - Costs 10 points, contains 1 "sink in a box" and 1 shower fixture.

![image](https://user-images.githubusercontent.com/8988254/29009326-5d8679ee-7af0-11e7-88af-d817ac5630b3.png)

Mirrors can now be unfastened from walls with a screwdriver, giving you a mountable mirror item.
- Shattered mirrors return a glass shard instead of a mirror.

Mirrors can be produced in the autolathe using just glass.

Fixes #8082

:cl:
rscadd: Toilets, sinks, showers, urinals, and mirrors can now be moved and constructed/purchased.
tweak: Boiling hot showers now generate steam faster, and freezing cold ones will remove steam faster.
tweak: Toolspeed now applies to adjusting the temperature on showers.
bugfix: Water tank crates no longer can generate cargo points in excess of their cost normally.
tweak: Showers no longer apply water while off and will re-apply water while you remain under a running one.
/:cl:

taukausanake commented:
That rotated sink hurts my eyes. I can't think of any better way for it, unless you do the thing the toilet is doing now and appear in front of human mobs.

Or, _you know_, NOT PUT SINKS FACING NORTH. Really is an eyesore and I hate it

outbools commented:
But this is what everything facing SOUTH does

Citinited commented:
Yeah, the perspective is really weird for rotated sinks. Probably something that could do with respriting, really... (not that anyone wants to do that 😄)

FalseIncarnate commented:
Sam, you mean north. South is the standard default dir and looks fine, north is flipped. If someone else wants to sprite it, be my guest. It seems strange to not allow a single direction though, which is actually a small factor of why I made kitchen sinks and urinals a single direction (they didn't have NORTH sprites and their EAST / WEST sprites were bad too). It essentially was "all or nothing" when it comes to rotating

Fox-McCloud commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +525,45 @@
 
 	if(!(istype(O)))
 		return
+	if(iswrench(O) && !istype(src, /obj/structure/sink/puddle))		//puddles laugh at your attempt to disconnect them. you can't remove a puddle from the ground with a wrench!
```
`istype(src, /obj/structure/sink/puddle))`
`istype(src)`

*scream

Come on, do this properly.

FalseIncarnate replied:
How would you prefer, a `can_move` and `can_rotate` set of vars? This check is so you can't move puddles which are a subtype of sink

tigercat2000 replied:
Yes, you should make those variables.

Fox-McCloud commented:
I do kinda question the legitimacy of being able to move around infinite water dispensers.

FalseIncarnate commented:
Honestly, infinite water really isn't that bad, most jobs that make use of water regularly typically already have a sink anyways, or easy access to a public one.

SamHPurp commented:
@FalseIncarnate uhh yeah. You know which one I meant.

Without the sprite re-done I don't think I could support it... the 'seems silly to have it in 3 directions but not the fourth is preferable to how that looks in the 4th right now. Sorry...

taukausanake commented:
I would really prefer we have another pipe system for water. Sounds like something Europa would have actually but another pipe system is maybe something no one wants ever. I do feel in the future water should be a finite resource but I don't know any means to making it one

SamHPurp commented:
@taukausanake oh god, that sounds awful. Having a water piping system.

Another thing for another server.... not for us.

Fox-McCloud commented:
TG was going to add in a water system for their server, but it got kinda sidelined.

It didn't need pipes though; it was semi-magical.

taukausanake commented:
Tell us a story @Fox-McCloud. I don't understand how it can be semi-magical

FalseIncarnate commented:
@taukausanake Once upon a time, there was a PR to add a plumbing system to TG. Everyone argued if it should belong to the janitor or engineering until it died. The end.

Nah, it was this magical tank thingy that you could add a reagent to and then all the sinks on the station would dispense that reagent. Literally let you make sinks of beer. There were a lot of arguments whether it should belong to the janitor or engineering in terms of where this tank was and who was responsible for it, plus probably some balance concerns, and eventually the PR was closed. You can read all about it here: https://github.com/tgstation/tgstation/pull/25337

taukausanake commented:
> Literally let you make sinks of beer

omg that sounds _amazing_ why would anyone want to close something like that now I want reagent pipes even more fuck dealing with another hell of a pipe system **it's worth it**

FalseIncarnate commented:
It didn't use pipes, it used BLUESPACE. Because why not?

taukausanake commented:
Only when we implement bluespace atmos pipes will I ever support a thing like that

FalseIncarnate commented:
Added the ability to detach and move showers (they turn into a mountable item like a light fixture)

Made a new north sink sprite.
![image](https://user-images.githubusercontent.com/8988254/28497208-9b43e4f8-6f4f-11e7-8320-b3a465e53bb2.png)

Oh yea, it's in front of you too. Suck on that perspective.

Updated original post with new picture and also to include the shower stuff. Still got a bit more before the WIP can come off this.

Kyep commented:
Could you make it possible to wrench/unanchor mirrors? Or just deconstruct them / construct new ones with glass? I ask because I've seen multiple ahelps the last few days to ask admins to delete them.

FalseIncarnate commented:
@Kyep Soon:tm:

Also going to be making mirrors craftable with glass (though wonder if that should be autolathe or just standard crafting) and the bathroom fixtures orderable via cargo crates.

Deciding on the costs and whether to be each as an individual crate or as bundles (a sink and a shower for the "hygiene" crate and a toilet and urinal for the "relief" crate)...

Leaning towards 10 cargo points for the cost per crate, using those two bundle ideas above. Feedback would be appreciated.

MoreRobustThanYou commented:
just make water pipes already

Citinited commented:
How does adding a huge system of water pipes, merely for the incredibly minor sink and shower objects, make the game and codebase better?

MoreRobustThanYou commented:
@Citinited muh immersion. also blowing people up with hot steam.

Citinited commented:
![](http://images.amcnetworks.com/ifc.com/wp-content/uploads/2016/03/steam.gif)

I guess, but it's still a major addition for such a minor thing...

taukausanake commented:
@Citinited Well it depends how it's implemented. That beer sink got me excited for the possibilities

MoreRobustThanYou commented:
*ahem* pumping CLF3 via plumbing

FalseIncarnate commented:
Toilets, urinals, sinks, and mirrors now have items you can use to build them with.

You can find toilet in a box and urinal in a box in the Urination Station Crate, and can find sink in a box and shower fixture in the Hygiene Station Crate. Both crates cost 10 cargo points each.

Mirrors can be produced in the autolathe from glass, and can be unfastened from walls with a screwdriver. Shattered mirrors drop a glass shard when unfastened instead of a mountable mirror.

The metal sinks (found in the kitchen, old bar, and perma-brig) cannot be produced or purchased from cargo. As mentioned earlier, they can be unfastened and moved.

FalseIncarnate commented:
Closed and re-opened to kick TRAVIS into trying again, failure was caused by a cancelled compile job on JUST metastation.

tigercat2000 commented on code/datums/supplypacks.dm:
```diff
@@ -1646,6 +1646,26 @@ var/list/all_supply_groups = list(supply_emergency,supply_security,supply_engine
+///////////// Bathroom Fixtures
+
+/datum/supply_packs/misc/toilet
+	name = "Urination Station Crate"
```
Urination Station? Do we *have* to do the rhyming?

taukausanake replied:
I like this. It's super cringe worthy and should stay

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -47,12 +47,62 @@
 
 /obj/structure/toilet/attackby(obj/item/I, mob/living/user, params)
+	if(iswrench(I))
+		if(anchored)
```
This is pretty messy. It'd probably be better formatted as

```dm
var/choices = list()
if(cistern)
	choices += "Stash [I]"

if(anchored)
	choices += "Disconnect"
else
	choices += "Connect"
	choices += "Rotate"

var/response = input(user, "What do you want to do?", "[src]") as null|anything in choices
if(!response) // User cancelled
	return

switch(response)
	if("Connect")
		if(anchored)
			return

		user.visible_message("<span class='notice'>[user] begins connecting [src]...</span>", "<span class='notice'>You begin connecting [src]...</span>")
		if(do_after(user, 40 * I.toolspeed, target = src))
			if(!loc || anchored)
				return
			user.visible_message("<span class='notice'>[user] connects [src]!</span>", "<span class='notice'>You connect [src]!</span>")
			anchored = TRUE
			update_icon()

	if("Disconnect")
		if(!anchored)
			return

		user.visible_message("<span class='notice'>[user] begins disconnecting [src]...</span>", "<span class='notice'>You begin disconnecting [src]...</span>")
		if(do_after(user, 40 * I.toolspeed, target = src))
			if(!loc || !anchored)
				return
			user.visible_message("<span class='notice'>[user] disconnects [src]!</span>", "<span class='notice'>You disconnect [src]!</span>")
			anchored = FALSE
			update_icon()

	if("Rotate")
		if(anchored)
			return
		var/list/choices = list("North" = NORTH, "East" = EAST, "South" = SOUTH, "West" = WEST)
		var/selected = input(user, "Select a direction for the connector.", "Connector Direction") as null|anything in choices
		if(!selected)
			return
		dir = choices[selected]
		update_icon()

	if("Stash")
		if(cistern)
			stash_goods(I, user)
```

FalseIncarnate replied:
That is quite frankly a much better way to handle it that I didn't think of at the time. Though, in complete honesty, I also didn't think I needed to make various exceptions when I began it either.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -141,6 +197,27 @@
+			if(do_after(user, 40 * I.toolspeed, target = src))
+				if(!loc || !anchored)
+					return
+				user.visible_message("[user] disconnected [src]!", "<span class='notice'>You disconnected [src]!</span>")
```
Present continuous tense messages are generally preferred to past tense messages. So this should be `[user] disconnects [src]!` and `You disconnect [src]!`.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -141,6 +197,27 @@
+			if(do_after(user, 40 * I.toolspeed, target = src))
+				if(!loc || anchored)
+					return
+				user.visible_message("[user] connected [src]!", "<span class='notice'>You connected [src]!</span>")
```
... and this should be `[user] connects [src]!` and `You connect [src]!`.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -141,6 +197,27 @@
 /obj/structure/urinal/attackby(obj/item/I, mob/user, params)
+	if(iswrench(I))
+		if(anchored)
+			user.visible_message("[user] begins disconnecting [src]...", "<span class='notice'>You begin to disconnect [src]...</span>")
```
All of these visible messages are missing span classes in the first message. Most of them should be notice.

FalseIncarnate replied:
Not arguing to keep it as such, but do want to point out a LOT of things in the code are missing spans for these kinds of messages. Even the message when you completed removing the lid from the toilet cistern prior to this PR was like this, as is also the case for things like pipes.

I'll still fix these messages, but at some point we'll probably want to go through and fix the other issues to avoid accidental referencing to improper styling.

tigercat2000 replied:
Pretty much nothing should be using bare black text, there are span classes that fit for pretty much every kind of message.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -173,6 +251,24 @@
+
+/obj/machinery/shower/Destroy()
+	if(mymist)
+		qdel(mymist)
```
Use `QDEL_NULL(mymist)`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -209,33 +305,57 @@
+			return
+		var/obj/item/weapon/weldingtool/WT = I
+		if(WT.isOn())
+			user.visible_message("[user] begins to cut [src] loose.", "<span class='notice'>You begin to cut [src] loose.</span>")
```
More missing spans.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -398,7 +518,9 @@
 		return
-
+	if(!anchored)
+		to_chat(user, "<span class='warning'>This [src] isn't connected, wrench it into position first!</span>")
```
Get rid of `This`, or you'll end up with `This the shower isn't connected`.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +567,45 @@
+		if(anchored)
+			var/response = alert(user, "Would you like to wash [W] or disconnect [src]?", "Wrench", "Wash", "Disconnect")
+			if(response == "Disconnect")
+				user.visible_message("[user] is disconnecting [src].", "<span class='notice'>You are disconnecting [src]...</span>")
```
Tenses and span classes are fucked up again.
`"<span class='notice'>[user] starts disconnecting [src]...</span>", "<span class='notice'>You start disconnecting [src]...</span>`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +567,45 @@
+				if(do_after(user, 40 * W.toolspeed, target = src))
+					if(!loc || !anchored)
+						return
+					user.visible_message("[user] disconnected [src]!", "<span class='notice'>You disconnected [src]!</span>")
```
`"<span class='notice'>[user] disconnects [src]!</span>", "<span class='notice'>You disconnect [src]!</span>"`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +567,45 @@
+		else
+			var/response
+			if(istype(src, /obj/structure/sink/kitchen))
+				response = "Connect"	//Kitchen sinks can't be rotated because they only have one directional state (SOUTH)
```
Yeaaah, you desperately need those `can_rotate` and `can_anchor` variables.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +567,45 @@
+			else
+				response = alert(user, "Would you like to rotate or connect [src]?", "Wrench", "Rotate", "Connect", "Cancel")
+			if(response == "Connect")
+				user.visible_message("[user] is connecting [src].", "<span class='notice'>You are connecting [src]...</span>")
```
Tenses and spans.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -445,6 +567,45 @@
+				if(do_after(user, 40 * W.toolspeed, target = src))
+					if(!loc || anchored)
+						return
+					user.visible_message("[user] connected [src]!", "<span class='notice'>You connected [src]!</span>")
```
Tenses and spans.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -453,6 +614,30 @@
+		pixel_x = 0
+		pixel_y = 0
+	else
+		if(istype(src, /obj/structure/sink/puddle))
```
Dear lord.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -471,3 +656,77 @@
+	if(proximity_flag != 1) //if we aren't next to the wall
+		return
+	if(!(get_dir(on_wall,user) in cardinal))
+		to_chat(user, "<span class='rose'>You need to be standing next to a wall to place \the [src].</span>")
```
I didn't even know `rose` is a span class. Use warning.

FalseIncarnate replied:
That was copied right from mountable code. I didn't realize rose was a span either, but I'll change it for everything to maintain consistency.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -471,3 +656,77 @@
+	if(result in T.contents)
+		to_chat(user, "<span class='warning'>There's already \an [result_name] here.</span>")
+		return
+	visible_message("[user] begins assembling a new [result_name].", "You begin assembling a new [result_name].")
```
These are both missing spans.

FalseIncarnate commented on code/datums/supplypacks.dm:
```diff
@@ -1646,6 +1646,26 @@ var/list/all_supply_groups = list(supply_emergency,supply_security,supply_engine
+///////////// Bathroom Fixtures
+
+/datum/supply_packs/misc/toilet
+	name = "Urination Station Crate"
```
No, but I couldn't think of a better crate name since "Relief crate" (the original name I came up with) didn't feel like it was obvious what it contained. Open to suggestions.

FalseIncarnate commented:
Cleaned up the code based on feedback from @tigercat2000 

Still have not renamed the Urination Station crate. I'm still struggling to come up with a better name that still communicates that it has a toilet and a urinal in the crate. That can always be renamed in the future though.

tigercat2000 commented on code/game/objects/structures/mirror.dm:
```diff
@@ -38,7 +53,19 @@
-/obj/structure/mirror/attackby(obj/item/I as obj, mob/living/user as mob, params)
+/obj/structure/mirror/attackby(obj/item/I, mob/living/user, params)
+	if(isscrewdriver(I))
+		visible_message("[user] begins to unfasten the [src].", "You begin to unfasten the [name].")
```
Missing spans.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -47,12 +47,61 @@
+			if("Stash")
+				stash_goods(I, user)
+			if("Disconnect")
+				user.visible_message("<span class='notice'>[user] is disconnecting [src].</span>", "<span class='notice'>You are disconnecting [src]...</span>")
```
The tenses here got a bit mixed up. `"<span class='notice'>[user] starts disconnecting [src]...</span>", "<span class='notice'>You start disconnecting [src]...</span>"`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -47,12 +47,61 @@
+					anchored = 0
+					update_icon()
+			if("Connect")
+				user.visible_message("<span class='notice'>[user] is connecting [src].</span>", "<span class='notice'>You are connecting [src]...</span>")
```
Tenses. `"<span class='notice'>[user] starts connecting [src]...</span>", "<span class='notice'>You start connecting [src]...</span>"`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -106,20 +155,25 @@
+		to_chat(user, "<span class='warning'>The cistern is full!</span>")
+		return
+	if(!user.drop_item())
+		to_chat(user, "<span class='warning'>\The [I] is stuck to your hand, you cannot put it in the cistern!</span>")
```
No `\The` please, it's redundant unless you are referring to `[I.name]`.

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -446,21 +569,100 @@
+				if(wateract)
+					W.water_act(20,310.15,src)
+			if("Disconnect")
+				user.visible_message("<span class='notice'>[user] is disconnecting [src].</span>", "<span class='notice'>You are disconnecting [src]...</span>")
```
Tenses. `"<span class='notice'>[user] starts disconnecting [src]...</span>", "<span class='notice'>You start disconnecting [src]...</span>"`

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -446,21 +569,100 @@
+					anchored = 0
+					update_icon()
+			if("Connect")
+				user.visible_message("<span class='notice'>[user] is connecting [src].</span>", "<span class='notice'>You are connecting [src]...</span>")
```
Tenses. `"<span class='notice'>[user] starts connecting [src]...</span>", "<span class='notice'>You start connecting [src]...</span>"`

FalseIncarnate commented:
This is ready for more review/merging if no more issues are raised. The name of the crate still stands until a better one is found.

FalseIncarnate commented:
Added a tweak to showers to fix them applying water to mobs while off while additionally making them re-wash mobs who remain under an active shower. Now you can murder the clown in the showers and get their banana-scented blood rinsed off without moving!

FalseIncarnate commented:
Urination Station Crate has been renamed to Lavatory Crate. Contents and price unchanged. Not going to update the original post's screenshot since that's unnecessary effort for a minor change.

tigercat2000 commented on code/game/objects/structures/mirror.dm:
```diff
@@ -38,7 +53,19 @@
-/obj/structure/mirror/attackby(obj/item/I as obj, mob/living/user as mob, params)
+/obj/structure/mirror/attackby(obj/item/I, mob/living/user, params)
+	if(isscrewdriver(I))
+		visible_message("<span class='notice'>[user] begins to unfasten the [src].</span>", "<span class='notice'>You begin to unfasten the [name].</span>")
```
Drop `the` before `[src]`

tigercat2000 replied:
Also, change the second message from `[name]` to `[src]` and drop `the` there as well

tigercat2000 replied:
None of these visual messages actually work, by the way- You're calling `visible_message` on the mirror, you would have to call `user.visible_message` in order to use the second argument

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -446,21 +572,100 @@
+					update_icon()
+			if("Rotate")
+				var/list/dir_choices = list("North" = NORTH, "East" = EAST, "South" = SOUTH, "West" = WEST)
+				var/selected = input(user,"Select a direction for the connector.", "Connector Direction") in dir_choices
```
Space after `user,`

FalseIncarnate replied:
this was fixed. git needs to git gud

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -471,3 +676,77 @@
+		return
+	if(proximity_flag != 1) //if we aren't next to the wall
+		return
+	if(!(get_dir(on_wall,user) in cardinal))
```
Space after `on_wall, `

tigercat2000 commented on code/game/objects/structures/watercloset.dm:
```diff
@@ -471,3 +676,77 @@
+	if(result in T.contents)
+		to_chat(user, "<span class='warning'>There's already \an [result_name] here.</span>")
+		return
+	visible_message("<span class='notice'>[user] begins assembling a new [result_name].</span>", "<span class='notice'>You begin assembling a new [result_name].</span>")
```
More visible_messages that use the second argument but aren't called on user

DesolateG commented:
Time to install all the toilets on the bridge
